### DISCUSSION TITLE
OPENSSL 3.x compatible providers information

### DISCUSSION THREAD

**Discussion by Rekha-2017** [openssl/openssl#23227] [Q&A] [answered]
Hi Team,

While working on the upgrade to OpenSSL 3.x version we are facing a query regarding provider compatibility. Out of the standard providers available ( Default, FIPS, Base, legacy, NULL) which of them can be loaded at a single time without having any issues?

I am sure that FIPS and BASE can be loaded together while using the FIPS enable mode. 
For non-FIPS mode, can we use default and legacy together? The link given below says that we can use default and legacy at a time.
https://github.com/openssl/openssl/blob/master/README-PROVIDERS.md#the-default-provider

I would like to know about other combinations that are allowed from this set.

Thanks,
Rekha

**Comment by levitte** [comment]
You should be able to load any combination of providers and use the implementations they provide (given appropriate property queries if necessary).  That's one of the criteria we had when we put together the architecture for how providers were going to work.

**Comment by slontis** [comment, marked as answer]
Each provider just implements algorithms.. So you load whatever providers you need the algorithms for.
You can use  any of the providers at the same time, however if an algorithm is available in multiple providers then there is no guarantee which provider it comes from UNLESS you specify a property query which can select the required algorithm.
e.g.
provider=fips
fips=yes

Property queries can be set per library context using EVP_default_properties_enable_fips() or EVP_set_default_properties().
These properties are applied any time a call such as EVP_MD_fetch is used.. but can be overridden by passing another property query value to any fetch. So you could use EVP_default_properties_enable_fips() which will chose the fips provider everywhere using EVP_MD_fetch(libctx, fipsalg, NULL) but then override it when needed e.g.  EVP_MD_fetch(libctx, "MD5", "fips=no");

Base is built into the default provider so you dont require both of these in the same libctx.
null should just be used on its own in a libctx, since its purpose was to have the default NULL libctx use this.
Base + FIPS are needed together since FIPS doesnt know how to encode/decode, but if you use FIPS + default then you dont need base.
legacy can be used by itself, with default OR FIPS+base or default+FIPS

**Comment by Rekha-2017** [comment]
Thanks, levitte, slontis for confirming the usage of different providers.  I will start testing by loading the required combination of providers  Though we are using a programmatic approach to load the providers,  we should be able to set the required settings.
